### PR TITLE
Arch Migrator: Use cross-compilation for Linux aarch64 && ppc64le

### PR DESCRIPTION
**Pull request by martin-g**
The PR is based on https://github.com/conda-forge/gromacs-feedstock/pull/43 but [uses](https://github.com/conda-forge/gromacs-feedstock/pull/49/commits/c173e46bcdb7d067e640b23ce2546e09454412bc) cross-compilation for Linux ARM64 and PPC64LE because TravisCI is very unreliable!

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by martin-g** [comment]
Arch Migrator: Use cross-compilation for Linux aarch64 && ppc64le

**Comment by martin-g** [comment]
@conda-forge-admin, please rerender

**Comment by martin-g** [comment]
@conda-forge-admin, please rerender

**Comment by mabraham** [comment]
Probably should bump the build number when it works. I have no idea how the cross compilation works, but if the tests pass then LGTM!

**Comment by martin-g** [comment]
The Linux ARM64 build fails with:
```
-- Performing Test CXX_COMPILE_WORKS_WITHOUT_SPECIAL_FLAGS
CMake Error at cmake/gmxManageSimd.cmake:50 (message):
  Cannot find SSE2 compiler flag.  Use a newer compiler, or disable SIMD
  support (slow).
```

**Comment by martin-g** [comment]
Does Gromacs support ppc64le ?
What value for `simdflavors` should be used for it ?
https://github.com/conda-forge/gromacs-feedstock/blob/00543141c1f0d3904eee62a3450eb869e8aae7b6/recipe/build.sh#L9C5-L9C16

**Comment by mabraham** [comment]
> Does Gromacs support ppc64le ? What value for `simdflavors` should be used for it ? https://github.com/conda-forge/gromacs-feedstock/blob/00543141c1f0d3904eee62a3450eb869e8aae7b6/recipe/build.sh#L9C5-L9C16

Yes I think so. Use `IBM_VSX`. See https://manual.gromacs.org/current/install-guide/index.html#simd-support

**Comment by martin-g** [comment]
It builds fine now for Linux ARM64 but some test(s) fail - https://dev.azure.com/conda-forge/84710dde-1620-425b-80d0-4cf5baca359d/_apis/build/builds/940372/logs/82

**Comment by martin-g** [comment]
The problem is with the [hardware_info_command](https://github.com/conda-forge/gromacs-feedstock/blob/00543141c1f0d3904eee62a3450eb869e8aae7b6/recipe/build.sh#L155):
```
++ cat /proc/cpuinfo '|' grep -m1 ''\''^flags'\'''
2024-05-22T11:32:25.7458593Z cat: invalid option -- 'm'
```
Currently https://github.com/conda-forge/gromacs-feedstock/pull/49/files#diff-d7075654874cb08007a21aaab3ecd4b3453a9087e7505d034d548b8938b599bcR176 uses only `uname -m` to detect the CPU architecture but since this is a cross-compilation it thinks it is `x86_64` and tries to detect the AVX flavors ...

I am not sure why `cat /proc/cpuinfo | grep -m1 ...` says that `cat` does not have `-m` argument ...

**Comment by mabraham** [comment]
> The problem is with the [hardware_info_command](https://github.com/conda-forge/gromacs-feedstock/blob/00543141c1f0d3904eee62a3450eb869e8aae7b6/recipe/build.sh#L155):
> 
> ```
> ++ cat /proc/cpuinfo '|' grep -m1 ''\''^flags'\'''
> 2024-05-22T11:32:25.7458593Z cat: invalid option -- 'm'
> ```
> 
> Currently https://github.com/conda-forge/gromacs-feedstock/pull/49/files#diff-d7075654874cb08007a21aaab3ecd4b3453a9087e7505d034d548b8938b599bcR176 uses only `uname -m` to detect the CPU architecture but since this is a cross-compilation it thinks it is `x86_64` and tries to detect the AVX flavors ...

That script runs when a user tries to use GROMACS. In general, a platform might support multiple SIMD flavours for which we have pre-compiled different binaries. This script decides which one to use. As such, it isn't going to work properly in a cross-compilation setup, but we don't need it to do that because we anyway can't run anything (including tests) without having the target hardware available. (I assume you saw the error after the build and install completed when something tried to run a test.)

> I am not sure why `cat /proc/cpuinfo | grep -m1 ...` says that `cat` does not have `-m` argument ...

Me neither

**Comment by mabraham** [comment]
> It builds fine now for Linux ARM64 but some test(s) fail - https://dev.azure.com/conda-forge/84710dde-1620-425b-80d0-4cf5baca359d/_apis/build/builds/940372/logs/82

Link doesn't load anything for me

File: recipe/build.sh [hunk 1]
```diff
@@ -164,10 +174,14 @@ esac
...
-  if [[ "\$uname" == "arm64" ]]; then
-    # Assume ARM Mac
+  if [[ "\$uname" == "arm" || "\$uname" == "aarch64" ]]; then
+    # Assume ARM Mac/Linux
     test -d "${PREFIX}/bin.ARM_NEON_ASIMD" && \
       simdflavor='ARM_NEON_ASIMD'
+  elif [[ "\$uname" == "ppc64le" ]]; then
+    # Assume PowerPC Linux
+    test -d "${PREFIX}/bin.IBM_VSX" && \
+      simdflavor='ARM_NEON_ASIMD'
```
Comment (mabraham) [inline recipe/build.sh:184]:
```suggestion
      simdflavor='IBM_VSX'
```

**Comment by martin-g** [comment]
> and install completed when something tried to run a test

Yes! I think it happens before the test phase [starts](https://github.com/conda-forge/gromacs-feedstock/pull/49/files#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9aR57-R65)
I wonder how the `osx-arm64`  tests worked until now since they also [use](https://github.com/conda-forge/gromacs-feedstock/pull/49/files#diff-478786365dd93f740eb520c2faa03d7a0623273663b75472272c8ef94297bbe2R4) cross-compilation.

**Comment by martin-g** [comment]
I think the tcsh script would break for the cross-compiled archs - https://github.com/conda-forge/gromacs-feedstock/pull/49/files#diff-d7075654874cb08007a21aaab3ecd4b3453a9087e7505d034d548b8938b599bcR215
This evaluates eagerly the cpu architecture at build time and it is wrong later at run time.

**Comment by martin-g** [comment]
@conda-forge-admin, please rerender

**Comment by martin-g** [comment]
Nice! The build is green now!
If #50 is merged first then I will rebase this PR!
But this PR also includes the improvements in #50, so #50 could be closed if you are OK to merge two "features" at once.

**Comment by mabraham** [comment]
I merged #50 so we can see clearly in the history what happened when if we later run into issues

**Review by mabraham — APPROVED** [review]
Great work!

**Comment by martin-g** [comment]
@conda-forge-admin, please rerender

**Comment by martin-g** [comment]
@conda-forge-admin, please rerender

**Comment by martin-g** [comment]
@conda-forge-admin, please rerender

**Comment by martin-g** [comment]
The 4 failing PR checks fail with `failed with error: [Errno 28] No space left on device` ...
Any ideas what could be done to overcome this ?

**Comment by martin-g** [comment]
@conda-forge-admin, please rerender

**Comment by martin-g** [comment]
Progress! Now the build fails with 
```
aarch64-conda-linux-gnu/bin/ld: /usr/local/cuda/lib64/libcudart_static.a: error adding symbols: file in wrong format
```

**Comment by martin-g** [comment]
> @martin-g It seems you are very close here! Can you finish this PR ?

I very much would like to but I am not sure why it breaks ...
It fails only for CUDA 11.8 and `double=no`. 
I've found a similar issue:
- https://forums.developer.nvidia.com/t/solved-libcudart-so-error-adding-symbols-file-in-wrong-format/71084/7
but I don't see a solution ...

**Comment by martin-g** [comment]
@conda-forge-admin, please ping team

**Comment by martin-g** [comment]
@conda-forge-admin, please rerender

**Comment by mabraham** [comment]
Skipping seems like the practical solution. Few will use GROMACS with conda on ppc and need cuda

**Comment by martin-g** [comment]
Thanks for the feedback, @mabraham !
More opinions ?
I hope the PR will be merged soon!

**Review by mabraham — APPROVED** [review]

**Pull request merged**